compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs

```diff
@@ -801,73 +821,220 @@ pub(crate) fn collect_global(
-        fm.add_file_with_source(sub_dir_nr_path.as_path(), "fn foo() {}".to_string());
-
-        // First check for the sub_dir.nr file and add it to the FileManager
-        let sub_dir_file_id = find_module(&fm, file_id, sub_dir_name).unwrap();
-
-        // Now check for files in it's subdirectory
+    fn finds_module_in_sibling_dir_for_regular_name() {
+        let dir = PathBuf::new();
+        let mut fm = FileManager::new(&dir);
+
+        // Create this tree structure:
+        // - sub_dir.nr
+        // - sub_dir/foo.nr
+        let sub_dir_file_id = add_file(&mut fm, &dir.join("sub_dir.nr"));
+        add_file(&mut fm, &dir.join("sub_dir").join("foo.nr"));
+
+        // `mod foo` from `sub_dir.nr` should find `sub_dir/foo.nr`
         find_module(&fm, sub_dir_file_id, "foo").unwrap();
     }
+
+    #[test]
+    fn finds_module_in_sibling_dir_mod_nr_for_regular_name() {
+        let dir = PathBuf::new();
+        let mut fm = FileManager::new(&dir);
+
+        // Create this tree structure:
+        // - sub_dir.nr
+        // - sub_dir/foo/mod.nr
+        let sub_dir_file_id = add_file(&mut fm, &dir.join("sub_dir.nr"));
+        add_file(&mut fm, &dir.join("sub_dir").join("foo").join("mod.nr"));
+
+        // `mod foo` from `sub_dir.nr` should find `sub_dir/foo.nr`
+        find_module(&fm, sub_dir_file_id, "foo").unwrap();
+    }
+
+    #[test]
+    fn finds_module_in_sibling_dir_for_special_name() {
+        let dir = PathBuf::new();
+        let mut fm = FileManager::new(&dir);
+
+        // Create this tree structure:
+        // - lib.nr
+        // - sub_dir.nr
+        // - sub_dir/foo.nr
+        let lib_file_id = add_file(&mut fm, &dir.join("lib.nr"));
+        add_file(&mut fm, &dir.join("sub_dir.nr"));
+        add_file(&mut fm, &dir.join("sub_dir").join("foo.nr"));
+
+        // `mod sub_dir` from `lib.nr` should find `sub_dir.nr`
+        let sub_dir_file_id = find_module(&fm, lib_file_id, "sub_dir").unwrap();
+
+        // `mod foo` from `sub_dir.nr` should find `sub_dir/foo.nr`
+        find_module(&fm, sub_dir_file_id, "foo").unwrap();
+    }
+
+    #[test]
+    fn finds_mod_dot_nr_for_special_name() {
+        let dir = PathBuf::new();
+        let mut fm = FileManager::new(&dir);
+
+        // Create this tree structure:
+        // - lib.nr
+        // - foo/mod.nr
+        let lib_file_id = add_file(&mut fm, &dir.join("lib.nr"));
+        add_file(&mut fm, &dir.join("foo").join("mod.nr"));
+
+        // Check that searching "foo" finds the mod.nr file
+        find_module(&fm, lib_file_id, "foo").unwrap();
+    }
+
+    #[test]
+    fn errors_mod_dot_nr_in_same_directory() {
+        let dir = PathBuf::new();
+        let mut fm = FileManager::new(&dir);
+
+        // Create this tree structure:
+        // - lib.nr
+        // - mod.nr
+        let lib_file_id = add_file(&mut fm, &dir.join("lib.nr"));
+        add_file(&mut fm, &dir.join("foo").join("mod.nr"));
```

Comment:
@TomAFrench for the given tree structure, the second line should be:

```rust
add_file(&mut fm, &dir.join("mod.nr"));
```

(without `.join("foo")`, because otherwise it's `foo/mod.nr`)

I think that's why this test is finding the file. If I remove `.join("foo")` the test correctly fails.

Reply:
Yep, you're right. My bad.